### ISSUE TITLE
Overlay implementation depends on ordering of rtree results

### ISSUE THREAD

**Issue by adriangb** [geopandas/geopandas#1337]
The current implementation of `tools.overlay` relies upon the ordering of the results from `rtree`. Specifically, changing [this line](https://github.com/adriangb/geopandas/blob/eba5dfa8300be350c7eec32e889bc70590bd3999/geopandas/tools/overlay.py#L30)  for instance:
```python3
sidx = bbox.apply(lambda x: list(spatial_index.intersection(x)))
```
to
```python3
sidx = bbox.apply(lambda x: sorted(list(spatial_index.intersection(x))))
```
breaks the tests.

Since I rtree does not promise any type of ordering (I think it depends on insertion order?), we should not rely on that for the result.

I see a couple of options here:
1) Modify tests so that they don't rely on the ordering and make it clear in the docs that ordering is non-deterministic.
2) Sort the results as above (could impact performance?) _and_ modify the tests. I looked into this but it looks like the tests rely on files exported from QGIS, I'm not sure what the best way to validate the results is.

**Comment by jorisvandenbossche** [comment]
To be clear: it can give indeterministic results in terms of ordering, or does this also give *wrong* results? (meaning, wrong geometries)

**Comment by martinfleis** [comment]
Why does order matter in this case?

**Comment by adriangb** [comment]
The results aren't _wrong_, just un-ordered. 
The only reason this really matters is that I was trying to implement this with Pygeos and these tests were failing because the spatial index results in Pygeos are ordered differently. For example:
```python
sidx = bbox.apply(lambda x: list(spatial_index.intersection(x)))
```
rtree:
```python
sidx
0    [1, 0, 2]
1       [0, 2]
```
pygeos:
```python
0    [0, 1, 2]
1       [0, 2]
```
So the results are both correct, but result in different row ordering for the resultant DataFrame in the case of `_overlay_intersection` and slightly different geometries in the case of `_overlay_difference`.

I think ideally the order of the results shouldn't depend on the spatial index implementation being used, unless we can explain clearly where that order comes from.

**Comment by jorisvandenbossche** [comment]
I tend to agree that it would be good to have a deterministic output order (now, in practice we probably have right now, since internally we always use rtree).

So sorting the indices returned by the spatial index might be fine?

> slightly different geometries in the case of _overlay_difference.

Can you explain this a bit more? So it does actually return different geometries? (and not only order of rows)

**Comment by jorisvandenbossche** [comment]
> Can you explain this a bit more? So it does actually return different geometries? (and not only order of rows)

Sorry, see now that you explained this in https://github.com/geopandas/geopandas/pull/1338#issuecomment-602624858. So it is about the order of the MultiPolygon parts
